DR-2515 - GCP Dataset View Data

snf2ye opened the pull request:
### Overview
This PR creates a new endpoint that retrieves data from a dataset's table for GCP backed datasets.  A follow-on ticket will address the same changes but for Azure backed datasets ([DR-2937](https://broadworkbench.atlassian.net/browse/DR-2937)). It also includes the following parameters: sort, sort direction, filter, limit and offset. We plan to use this endpoint to display data in the TDR UI and to allow users to view data that lives in TDR. It will also be used as part of the create snapshot workflow in the TDR UI.

### Code Changes
These changes heavily mirror the snapshot preview endpoint. Notable differences:
- We will also return the `datarepo-row-id` column for dataset tables
- We have made a naming change to refer to this endpoint as a "view data" endpoint rather than a "preview" endpoint since users can access the entire dataset with this endpoint, not just a preview. (see [DR-2961](https://broadworkbench.atlassian.net/browse/DR-2961) for ticket to make equivalent renaming changes for snapshots)
- For the actual BigQuery query to retrieve the data, we must prefix the dataset name with "datarepo_"

For reference, PRs from equivalent snapshot preview endpoint:
- [Original Preview Snapshot PR](https://github.com/DataBiosphere/jade-data-repo/pull/1131)
-  [Repeated Column Fix](https://github.com/DataBiosphere/jade-data-repo/pull/1153)
- [Sort results by column](https://github.com/DataBiosphere/jade-data-repo/pull/1276)
- [Filter Results](https://github.com/DataBiosphere/jade-data-repo/pull/1309)

### Tests
- Harnessed the tests from the PRs above to also test the dataset equivalent functionality
- Replaced as much of the direct BigQuery queries with calls to the new endpoint (this should hopefully speed up our integration tests!)
- Added a few checks in an integration test to assert permissions are correctly checked

[DR-2961]: https://broadworkbench.atlassian.net/browse/DR-2961?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DR-2937]: https://broadworkbench.atlassian.net/browse/DR-2937?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

snf2ye committed:
First pass - Mostly copying logic from snapshot preview endpoint for gcp

snf2ye committed:
Share more of the code between datasets and snapshots

snf2ye committed:
Update Connected test to instead use preview endpoint

remove test

snf2ye committed:
Share getData code in BigQueryPdao

snf2ye committed:
Add units tests for datasets controller lookupDatasetDataById endpoint

- This copy the tests from the search API preview snapshot endpoint
- Plus spotless changes

snf2ye committed:
Add connected test case for dataset data endpoint filtering

- Harness Connected testPreview to run the same tests for both snapshots and dataset's preview code
- Add sort parameter for snapshot preview endpoint
- Add failure case for dataset data endpoint
- Added connectedOperations.retrieveData() so that we could use the same code to run same tests against both snapshots and dataset

snf2ye committed:
Update assertTableCount test helper to use data view endpoint rather than directly querying BQ

update row count from long to int

spotless

snf2ye committed:
Include datarepo_row_id in get dataset data endpoint

- switch getRowIds from direct bq to preview endpoint
- Get unit tests passing

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetService.java:
```diff
@@ -439,6 +460,98 @@ public List<BulkLoadHistoryModel> getLoadHistory(
+      logger.warn("Snapshot {} inaccessible via SAM for {}", datasetId, userReq.getEmail(), iamEx);
+      error = iamEx.getMessage();
+    } finally {
+      if (!iamAuthorized) {
+        throw new ForbiddenException("Error accessing snapshot: see errorDetails", List.of(error));
+      }
```
The only way that `iamAuthorized` is set to true is on line 481. So you could get rid of `iamAuthorized` and move this `throw` into the `catch` clause. You can also get rid of `error` once that's done.

snf2ye committed:
Update test utility transformStringResults to use new endpoint

snf2ye committed:
Remove query and queryWithRetry from BigQueryFixtures

cleanup

Fix test

snf2ye committed:
Update permission check code and add tests

- Update permission check based on code review
- Add tests for permission check

nmalfroy commented on src/main/java/bio/terra/service/dataset/DatasetService.java:
```diff
@@ -439,6 +459,86 @@ public List<BulkLoadHistoryModel> getLoadHistory(
+                    new DatasetDataException(
+                        "No dataset table exists with the name: " + tableName));
+
+    if (!sort.equalsIgnoreCase(PDAO_ROW_ID_COLUMN)) {
```
Is it possible for sort to be null?

snf2ye replied:
I actually meant to go back and remove this condition. This is copied from the snapshot preview code and snapshot preview currently doesn't include the row id column in its results. 
But, for the underlying call, sort has a default value of "datarepo_row_id" in the open api spec, so should not be null.

snf2ye replied:
Nevermind! I still need this condition because the row id is not included in the list of columns. 
But, to your original question - since sort has a default value, we should be fine.

nmalfroy commented on src/main/java/bio/terra/service/dataset/DatasetService.java:
```diff
@@ -439,6 +459,86 @@ public List<BulkLoadHistoryModel> getLoadHistory(
+      }
+    } else if (cloudPlatformWrapper.isAzure()) {
+      throw new NotImplementedException(
+          "Azure datasets are not yet supported for the preview endpoint");
```
should we not call it the preview endpoint?
```suggestion
          "Azure datasets are not yet supported for the data endpoint");
```

nmalfroy commented on src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQueryPdao.java:
```diff
@@ -92,4 +100,86 @@ public static boolean tooManyDmlStatementsOutstanding(PdaoException ex) {
+
+    String columns = String.join(",", columnNames);
+    // Parse before querying because the where clause is user-provided
+    final String sql =
```
This might want to be in shared code since this is part of the request validation and is needed for both azure and GCP

snf2ye replied:
Ah, that's a great point. Thanks for pointing that out. I think I will leave this for the Azure PR so I can take a broader look at where I want to put that shared code.

nmalfroy reviewed (approved):
phew, this looks great!  My comment about pulling the validation of the sql can totally be done in a followon

snf2ye committed:
Update src/main/java/bio/terra/service/dataset/DatasetService.java

Co-authored-by: Nick Malfroy-Camine <nmalfroy@broadinstitute.org>

snf2ye committed:
Allow sorting by datarepo_row_id, fix lingering references to snapshot

snf2ye committed:
Add comment to remember to share this code in DR-2937

okotsopoulos commented on src/test/java/bio/terra/app/controller/DatasetsApiControllerTest.java:
```diff
@@ -0,0 +1,154 @@
+      throws Exception {
+    var list = List.of("hello", "world");
+    var result = new DatasetDataModel().result(List.copyOf(list));
+    // when(datasetService.retrieve(id)).thenReturn(new Dataset().id(id));
```
Did you mean to remove this?

snf2ye replied:
Yes, thanks!

okotsopoulos commented on src/test/java/bio/terra/common/fixtures/ConnectedOperations.java:
```diff
@@ -843,14 +887,67 @@ public MockHttpServletResponse retrieveSnapshotPreviewByIdRaw(
     return result.getResponse();
   }
 
+  public enum TDRResourceType {
```
Nit: I think we like to follow Google's style guide, which uses camel case in names including for acronyms.  Under that scheme, this would be `TdrResourceType`. https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

okotsopoulos commented on src/test/java/bio/terra/integration/BigQueryFixtures.java:
```diff
@@ -62,76 +55,6 @@ public static boolean datasetExists(BigQuery bigQuery, String projectId, String
     }
   }
 
-  public static TableResult query(String sql, BigQuery bigQuery) {
```
It brings me a LOT of joy to see these fixtures removed!

snf2ye replied:
Me too! Although, I haven't seen the test timing improvements I was hoping for, but I still think this is a better approach.

okotsopoulos reviewed (approved):
This looks good to me!

snf2ye committed:
doh, row id sin't in the list of column names

spotless

snf2ye committed:
Code review Changes: Remove lingering comment, Update name

snf2ye commented:
Last test run actually passed - the whitelist cleanup step failed. Re-running to get clean run.
